File: tasks/2024/final/gossip.cpp
// @check-accepted: examples NQsmall Nsmall no-limits
// NOTE: it is recommended to use this even if you don't understand the following code.

#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    int N; cin >> N;
    vector<int> P(N);
    for (int i = 0; i < N; ++i)
        cin >> P[i];
    
    vector<int> cycle(N, -1);
    vector<int> cycle_sz(N, 0);
    vector<int> cycle_pos(N, 0);

    for(int i = 0; i < N; ++i){
        int c = i;
        while(cycle[c] == -1){
            cycle[c] = i;
            c = P[c];
            cycle_pos[c] = cycle_sz[i]++;
        }
    }
    
    int Q; cin >> Q;
    int a,b;
    while(Q--){
        cin >> a >> b;
        if(cycle[a] != cycle[b]){
            cout << -1 << " ";
        }else{
            int c = cycle[a];
            int cs = cycle_sz[c];
            int d = (cycle_pos[a] - cycle_pos[b] + cs) % cs;
            cout << min(d, cs - d) << " ";
        }
    }

    cout << endl;

    return 0;
}

File: tasks/2024/final/gossip.py
#!/usr/bin/env python3
# @check-accepted: examples NQsmall Nsmall no-limits
# NOTE: it is recommended to use this even if you don't understand the following code.

import sys

# uncomment the two following lines if you want to read/write from files
# sys.stdin = open('input.txt')
# sys.stdout = open('output.txt', 'w')

N = int(input().strip())

P = list(map(int, input().strip().split()))

Q = int(input().strip())

A = [0 for i in range(Q)]
B = [0 for i in range(Q)]
for i in range(Q):
    A[i], B[i] = map(int, input().strip().split())

R = [0 for i in range(Q)]

cycle = [-1 for i in range(N)]
cycle_len = [0 for i in range(N)]
cycle_pos = [0 for i in range(N)]

for i in range(N):
    c = i
    while cycle[c] == -1:
        cycle[c] = i
        cycle_len[i] += 1
        cycle_pos[c] = cycle_len[i]
        c = P[c]

for i in range(Q):
    if cycle[A[i]] != cycle[B[i]]:
        R[i] = -1
    else:
        d = abs(cycle_pos[A[i]] - cycle_pos[B[i]])
        R[i] = min(d, cycle_len[cycle[A[i]]] - d)


for i in range(Q):
    print(R[i], end=' ')
print()

sys.stdout.close()
